lib/WebSocket.js

```diff
@@ -17,6 +17,43 @@ wsAppHealth.on('connection', async function connection(ws, request) {
   ws.on('close', () => clearInterval(healthChecker));
 });
 
+const wsAppLogs = new WebSocket.Server({ noServer: true});
+wsAppLogs.on('connection', async function connection(ws, request) {
+  const appId = Number(url.parse(request.url, true).query.appId);
+  const app = await App.findByPk(appId);
+  const managerNode = await DockerWrapper.getManagerNodeInstance();
+  const service = managerNode.getService(`${app.title}_web`);
+  const duplex = WebSocket.createWebSocketStream(ws, { encoding: 'utf8' });
+
+  const options = {
+    stdout: true,
+    stderr: true,
+    follow: true,
+  };
+
+  service.logs(options, function handler(err, stream) {
+    // I noticed when simply piping the stream to the WS client that weird
+    // artifacts would sometimes be included at the beginning of lines.
+    // See: https://github.com/moby/moby/issues/32794
+    //
+    // It turns out that Dockerode (or really, Docker Modem, which Dockerode uses
+    // internally) comes with a nice piece of functionality for handling this,
+    // called `demuxStream`
+    // More info here: https://github.com/apocas/dockerode/issues/259#issuecomment-242434163
+    service.modem.demuxStream(stream, duplex, duplex);
+
+    stream.on('end', function(){
+      console.log('wsAppLogs | stream.on end')
+      stream.destroy();
+    });
+
+    ws.on('close', () => {
+      console.log('wsAppLogs | ws.on close')
+      stream.destroy();
```

Comment:
Did you confirm that this fires when the user closes the browser or leaves the page? For some reason, it wasn't firing for me in the terminal ws (and so in the case of long-running commands like `bash` that don't exit on their own, I end up leaving a running container behind, which isn't great). Was thinking about doing something like this: https://github.com/websockets/ws#how-to-detect-and-close-broken-connections

Reply:
Yep! I tested closing the tab and/or the window, and that event fires on the server-side. The event also fires if you exit from the CLI using a <kbd>Ctrl</kbd> + <kbd>C</kbd>.